File: src/scanners/RetroarchSource.js
const fsp = require("fs/promises");
const path = require("path");
const fs = require("fs");

const Source = require("./Source.js");
const RetroarchGame = require("../games/RetroarchGame");

const USER_DIR = process.env["HOME"];

class RetroarchSource extends Source {

	static name = "Retroarch";
	static gameClass = RetroarchGame;

	preferCache = false;

	configPath = `${USER_DIR}/.config/retroarch/playlists`;

	constructor(preferCache = false) {
		super();
		this.preferCache = preferCache;
	}

	/**
	 * Get the paths to retroarch playlists.
	 * Found in $HOME/.config/retroarch/playlists
	 * @returns {string[]} - An array of playlist paths
	 * @private
	 */
	async _getPlaylistPaths() {

		let playlists = await fsp.readdir(this.configPath, { encoding: "utf-8", withFileTypes: true });
		playlists = playlists.filter(dirent=>dirent.isFile() && dirent.name.endsWith(".lpl"));
		playlists = playlists.map(dirent=>`${this.configPath}/${dirent.name}`);
		return playlists;

	}

	/**
	 * Get retroarch games from a playlist path.
	 * @param {string} playlistPath - Path to the playlist file to read from
	 * @returns {RetroarchGame[]} - An array of found games.
	 * @private
	 */
	async _getGamesFromPlaylist(playlistPath) {

		// Read the playlist file (it's JSON)
		const fileContents = await fsp.readFile(playlistPath, "utf-8");
		const playlist = JSON.parse(fileContents);

		// Get playlist console and default playlist core
		const PLAYLIST_DEFAULT_CORE_PATH = playlist.default_core_path;
		const PLAYLIST_PLATFORM = path.basename(playlistPath, ".lpl");

		// Build games from the given entries
		const games = [];
		for (const entry of playlist.items) {

			const gamePath = entry.path;
			if (!gamePath) {
				continue;
			}

			const gameIsInstalled = fs.existsSync(gamePath);
			let gameName = entry.label;
			let gameCorePath = entry.corePath;

			// Validate game data
			if (!gameName) {
				gameName = path.basename(gamePath);
			}
			if (!gameCorePath || gameCorePath === "DETECT") {
				gameCorePath = PLAYLIST_DEFAULT_CORE_PATH;
			}
			const game = new this.constructor.gameClass(
				gameName,
				gamePath,
				gameCorePath,
				PLAYLIST_PLATFORM,
				gameIsInstalled
			);
			if (game.name && game.path && game.corePath && game.platform) {
				games.push(game);
			} else {
				// Debug info
				// console.log(game.name, game.path, game.corePath, game.platform);
			}

		}
		return games;

	}

	/**
	 * Get all retroarch games
	 * @param {boolean} warn - Whether to display additional warnings
	 * @returns {RetroarchGame[]} - An array of found games
	 */
	async scan(warn = false) {


		// Get retroarch playlists
		let playlistPaths = [];
		try {
			playlistPaths = await this._getPlaylistPaths();
		} catch (error) {
			if (warn){
				console.warn(`Unable to get retroarch playlists : ${error}`);
			}
		}


		// Read playlists
		const games = [];
		for (const playlistPath of playlistPaths) {
			let playlistGames;
			try {
				playlistGames = await this._getGamesFromPlaylist(playlistPath);
			} catch (error) {
				if (warn){
					console.warn(`Unable to get retroarch games from ${playlistPath} : ${error}`);
				}
				playlistGames = undefined;
			}
			if (playlistGames) {
				games.push(...playlistGames);
			}
		}

		return games;

	}

}

module.exports = RetroarchSource;

File: gali/sources/retroarch_source.py
import os
import json
from pathlib import PurePath

from gali.games.retroarch_game import RetroarchGame, RetroarchFlatpakGame
from gali.sources.source import Source
from gali.utils.cfg_parser import CfgParser
from gali.utils.locations import HOME
from gali.sources.file_dependent_scannable import FileDependentScannable


class RetroarchSource(Source, FileDependentScannable):

    name: str = "Retroarch"
    game_class: type[RetroarchGame] = RetroarchGame
    config_path: str = f"{HOME}/.config/retroarch/retroarch.cfg"

    def get_config(self) -> CfgParser:
        config = CfgParser()
        config.read(self.config_path)
        return config

    def get_playlist_paths(self, config: CfgParser) -> tuple[str]:

        # Get playlist dir
        playlists_dir: str = config.get("playlist_directory", None)
        if playlists_dir is None:
            raise KeyError()

        # Handle "~" for user home
        playlists_dir = os.path.expanduser(playlists_dir)

        def filter_fn(dirent: os.DirEntry):
            if not dirent.is_file():
                return False
            if not dirent.name.endswith(".lpl"):
                return False
            return True

        # Get playlist paths
        iterator = os.scandir(playlists_dir)
        playlists = tuple(filter(filter_fn, iterator))
        return playlists

    def get_games(self, playlist_paths: tuple[str]) -> tuple[RetroarchGame]:

        games = []

        for playlist_path in playlist_paths:

            # Get playlist data
            try:
                file = open(playlist_path, encoding="utf-8-sig")
                playlist = json.load(file)
            except OSError:
                continue
            else:
                file.close()

            # Get playlist platform
            platform = PurePath(playlist_path).stem

            # Get default core for the playlist
            pl_core_path = playlist.get("default_core_path", "")
            if len(pl_core_path) == 0:  # Handle empty string
                pl_core_path = None

            # Get games
            items = playlist.get("items", list())
            for item in items:

                # Skip broken games
                name = item.get("label", None)
                game_path = item.get("path", None)
                core_path = item.get("core_path", pl_core_path)
                if (
                    name is None or
                    game_path is None or
                    core_path is None
                ):
                    continue

                # Build games
                game = self.game_class(
                    name=name,
                    game_path=game_path,
                    core_path=core_path,
                    platform=platform
                )
                games.append(game)

        return tuple(games)

    def scan(self) -> tuple[RetroarchGame]:
        config = self.get_config()
        playlist_paths = self.get_playlist_paths(config)
        games = self.get_games(playlist_paths)
        return games

    def get_precondition_file_path(self):
        return self.config_path


class RetroarchFlatpakSource(RetroarchSource):

    name: str = "Retroarch (Flatpak)"
    game_class: type[RetroarchFlatpakGame] = RetroarchFlatpakGame
    config_path: str = f"{HOME}/.var/app/org.libretro.RetroArch\
/config/retroarch/retroarch.cfg"
